File: rust/p3.rs
/// ### Largest prime factor
/// ## Problem 3
/// The prime factors of 13195 are 5, 7, 13 and 29.
///
/// What is the largest prime factor of the number 600851475143 ?

use std::collections::TreeSet;

fn factors(num: int) -> Option<(int, int)> {
  for divisor in range(2, num / 2) {
    if num % divisor == 0 {
      return Some((divisor, num / divisor))
    }
  }
  None
}

fn prime_factors(num: int) -> TreeSet<int> {
  let mut numerator = num;
  let mut primes = TreeSet::<int>::new();

  loop {
    match factors(numerator) {
      Some((a, b)) => {
        primes.insert(a);
        numerator = b;
      }
      None => {
        primes.insert(numerator);
        return primes
      }
    }
  }
}

fn main() {
  let num = 600851475143;
  println!("{}", prime_factors(num).iter().max().unwrap());
}

File: scala/p3.scala
/*
 * 10001st prime
 * Problem 7
 * By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.
 * What is the 10 001st prime number?
 */

import scala.math.sqrt

object Euler3 {
    val primes: Stream[Int] = 2 #:: Stream.from(3).filter({ p => primes.takeWhile(_ <= sqrt(p)).forall(p % _ != 0) })
    
    def factors(num: Long): List[Long] = {
        primes.takeWhile(_ <= sqrt(num)).find(num % _ == 0) match {
            case Some(p) => p :: factors(num / p)
            case None => List(num)
        }
    }

    def main(args: Array[String]) {
        println(factors(600851475143L).max)
    }
}
